### ISSUE TITLE
WebSocket client drops some incoming messages when running locally

### ISSUE THREAD

**Issue by rnbrady** [vercel/edge-runtime#983]
## Bug Report

A WebSocket client instantiated in the Edge runtime will randomly drop incoming messages from the WedSocket server. In my test every 10th or so incoming message is dropped, but it probably depends on timing and message length (edit: it's actually triggered by a websocket keepalive).

**Possible solution**

This appears to be caused by by upstream bug [nodejs/undici#2859](https://github.com/nodejs/undici/issues/2859) which has been fixed and released in [nodejs/undici@6.17](https://github.com/nodejs/undici/releases/tag/v6.17.0). 

Bumping Undici from 5 to 6 appears to be a breaking change so Dependabot [hasn't been able](https://github.com/vercel/edge-runtime/pull/947) to do it automatically. Possibily complicating matters further is that Undici v6 does not support Node v16.

**Additional context/screenshots**

In my application I'm connecting from a server-side Edge function to a Bitcoin indexer over WebSocket to subscribe to balance updates which I then send to the client as server-sent events (SSE). After an initial handshake I ping the indexer over WebSocket every second, and the pongs are received except that every 10th or so one gets dropped and a null message is delivered:

<img width="617" alt="Screenshot 2024-10-04 at 16 20 26" src="https://github.com/user-attachments/assets/3e4fe785-9663-454c-81a6-415f5c822f20">

With OpenAI recently annoucing their [Realtime API](https://openai.com/index/introducing-the-realtime-api/) based on WebSockets, other developers might soon run into this problem too.

tl;dr: please try and merge #947 🙏

**Comment by rnbrady** [comment]
On further investigation the problem occurs when the server sends a websocket ping message. See [rnbrady/vercel-edge-runtime-983](https://github.com/rnbrady/vercel-edge-runtime-983) to reproduce.

<img width="1015" alt="Screenshot 2024-10-05 at 23 13 22" src="https://github.com/user-attachments/assets/4dd4a72d-fb3c-45e5-91ee-ff50cd405519">

**Comment by rnbrady** [comment]
I am unable to reproduce this bug in deployments to Vercel.

**Comment by rnbrady** [comment]
This might be resolved by #546.

**Comment by Kikobeats** [comment]
Hello, this is going to be tricky to resolve since edge-runtime still supports Node.js v16, and to make this possible we are sticky to undici v5:

https://github.com/vercel/edge-runtime/blob/main/packages/primitives/package.json#L33

The fix for the undici websocket version was shipped under undici v6, which is dropping Node.js v16 support.

So in order to ship this we need to drop Node.js v16 support and upgrade (or drop) Undici , which is covered in https://github.com/vercel/edge-runtime/issues/546

**Comment by rnbrady** [comment]
Thanks for the reply @Kikobeats. I'm happy to wait for #546. 

I just looked into how Cloudflare Workers does `WebSocket` in local dev and it turns out they have [problems](https://community.cloudflare.com/t/writing-a-websocket-client-to-connect-to-remote-websocket-server-doesnt-work/494853) too, so I appreciate it's not an easy probem to solve.

**Comment by rnbrady** [comment]
Thanks @Kikobeats, I will try it out once it's marged and let you know. Appreciate your work on this. Dev prod disparity in Edge functions has been taken me on quite a ride lately, so this will be a nice fix to have. Appreciate your work.

**Comment by Kikobeats** [comment]
It's shipped starting from [Next.js v15.0.4-canary.37](https://github.com/vercel/next.js/releases/tag/v15.0.4-canary.37)

Closing but please tell me if it's working fine for you now 🙂

**Comment by rnbrady** [comment]
@Kikobeats I can confirm this is fixed. Tested against 15.0.4-canary.41. Thank you so much. This will allow me to significantly simplify my dev environment.
